modules/internal/ChapelIO.chpl

```diff
@@ -17,36 +17,310 @@
+.. code-block:: chapel
+
+  class IntPair {
+    var x: int;
+    var y: int;
+    proc readWriteThis(f) {
+      f <~> x <~> new ioLiteral(",") <~> y <~> new ioNewline();
+    }
+  }
+  var ip = new IntPair(17,2);
+  write(ip);
+  // prints out
+  // 17,2
+
+  delete ip;
+
+This example defines a only a writeThis method - so that there will be a
+function resolution error if the class NoRead is read.
+
+.. code-block:: chapel
+
+  class NoRead {
+    var x: int;
+    var y: int;
+    proc writeThis(f:Writer) {
+      f.writeln("hello");
+    }
+    // Note that no readThis function will be generated.
+  }
+  var nr = new NoRead();
+  write(nr);
+  // prints out
+  // hello
+
+  // Note that read(nr) will generate a compiler error.
+
+  delete nr;
+
+
+Generalized write and writeln
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The :class:`Writer` class contains no arguments and serves as a base class to
+allow user-defined classes to be written to.  If a class is defined to be a
+subclass of Writer, it must override the :class:`Writer.writePrimitive` method
+that any Chapel primitive type as an argument.
+
+The following code defines a subclass of :class:`Writer` that overrides the
+:proc:`Writer.writePrimitive` method to allow it to be written to.  It also
+overrides the ``writeThis`` method to override the default way that it is
+written.
+
+.. code-block:: chapel
+
+  // A demonstration writer class that filters
+  // values printed to it by only printing the first
+  // character.
+  class C: Writer {
+    // A variable to accumulate some writes.
+    var data: string;
+
+    // writePrimitive will be called when
+    // c.write(...) is called
+    proc writePrimitive(x) {
+      var s = x:string;
+      // only save the first letter
+      data += s.substring(1);
+    }
+
+    // writeThis will be called when
+    // e.g. stdout.write(c) is called
+    proc writeThis(x: Writer) {
+      x.write(data);
+    }
+  }
+
+  var c = new C();
+  // c.write will invoke c.writePrimitive
+  // with each numeric value in turn
+  c.write(41, 32, 23, 14);
+  // writeln is stdout.writeln and will invoke
+  // c.writeThis.
+  writeln(c);
+  // prints out:
+  // 4321
+
+  delete c;
+
+
+.. _default-write-and-read-methods:
+
+Default write and read Methods
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Default ``write`` methods are created for all types for which a user-defined
+``write`` method is not provided.  They have the following semantics:
+
+* for an array argument, outputs the elements of the array in row-major order
+  where rows are separated by line-feeds and blank lines are used to separate
+  other dimensions.
+* for a `domain` argument,  outputs the dimensions of the domain enclosed by
+  ``[`` and ``]``.
+* for a `range` argument, output the lower bound of the range followed by
+  ``..`` followed by the upper bound of the range.  If the stride of the range
+  is not one, the output is additionally followed by the word ``by`` followed
+  by the stride of the range.
```

Comment:
Instead of "followed by the stride of the range", I'd use "and by the stride of the range".